src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -61,33 +74,75 @@ new ::
   -> Mappend
   -> Run.RunFsPaths
   -> V.Vector (Run IO (FS.Handle h))
-  -> IO (Maybe (Merge IO (FS.Handle h)))
+  -> IO (Maybe (Merge IO h))
 new fs hbio mergeCaching alloc mergeLevel mergeMappend targetPaths runs = do
     -- no offset, no write buffer
     mreaders <- Readers.new fs hbio Readers.NoOffsetKey Nothing runs
     for mreaders $ \mergeReaders -> do
       -- calculate upper bounds based on input runs
       let numEntries = coerce (sum @V.Vector @Int) (fmap Run.runNumEntries runs)
       mergeBuilder <- Builder.new fs targetPaths numEntries alloc
-      return Merge {..}
+      mergeLastStepResult <- newMutVar $! MergeInProgress
+      return Merge {
+          mergeHasFS = fs
+        , mergeHasBlockIO = hbio
+        , ..
+        }
+
 
 -- | This function should be called when discarding a 'Merge' before it
 -- was done (i.e. returned 'MergeComplete'). This removes the incomplete files
 -- created for the new run so far and avoids leaking file handles.
 --
 -- Once it has been called, do not use the 'Merge' any more!
-close ::
-     HasFS IO h
-  -> HasBlockIO IO h
-  -> Merge IO (FS.Handle h)
-  -> IO ()
-close fs hbio Merge {..} = do
-    Builder.close fs mergeBuilder
-    Readers.close fs hbio mergeReaders
+close :: Merge IO h -> IO ()
+close Merge {..} = do
+    Builder.close mergeHasFS mergeBuilder
+    Readers.close mergeHasFS mergeHasBlockIO mergeReaders
+
+-- | Complete a 'Merge', returning a new 'Run' as the result of merging the
+-- input runs. This function will /not/ do any merging work if there is any
+-- remaining. That is, if not enough 'steps' were performed to exhaust the input
+-- 'Readers's, this function will throw an error.
+--
+-- Note: this function creates new 'Run' resources, so it is recommended to run
+-- this function with async exceptions masked. Otherwise, these resources can
+-- leak.
+complete :: Merge IO h -> IO (Run IO (FS.Handle h))
+complete Merge{..} = do
+    readMutVar mergeLastStepResult >>= \case
+      MergeInProgress -> error "complete: Merge is not yet completed!"
+      MergeComplete -> do
```

Comment:
Now that I think about it, we could make this API a bit safer by taking some `CompletedMerge` newtype argument that `steps` only returns in the case where the merge is complete.
Doesn't buy us a lot, unfortunately. We still need `mergeLastStepResult` for checking the opposite though (`steps` only being called on in-progress merges).

Reply:
Mmm, not clear to me either way.

Reply:
Hmm yes, though I think it would complicate the `MergeSchedule` code a bit. With scheduling, `complete` is not necessarily called immediately after `steps`. Well, that's if we follow the prototype precisely, but we might deviate from that once we get to actually implementing it. For now, I'll leave the code as is